### ISSUE TITLE
Unable to use subschemas for validation (fragment/anchor)

### ISSUE THREAD

**Issue by ssbarnea** [python-jsonschema/jsonschema#931]
Most JSON Schema validators allow use of a subschema to validate a file but I was not able to find a way to do that using python-jsonschema.

To example it, a subschema reference is something like `http://example.com/app.schema.json#/definitions/foo`. That URI tells the loader that the schema to be used is not at root of the JSON object, and instead is as a sublevel.

You can see a live example at https://github.com/SchemaStore/schemastore/blob/master/src/api/json/catalog.json#L222 or https://github.com/ansible/schemas#schema-bundle where you can see two subschema urls.

Keep in mind that that the subschema is not just a sub-json document as it self in can reference definitions from the entire file, so you cannot just handle the subdocument as a full schema.

Is there a way to do this with python-jsonschema?

PS. I know for sure that ajv works fine with subschemas, and also vscode json and yaml extensions. All of these can load and use schemas using the fragment URIs.

**Comment by MatthewNielsen27** [comment]
@ssbarnea I believe I'm running into a similar issue (albeit with subschemas defined in `$defs`). 

Like you mentioned:

> Keep in mind that that the subschema is not just a sub-json document as it self in can reference definitions from the entire file, so you cannot just handle the subdocument as a full schema.

One schema can even reference another's subschema (but not recommended) as outlined in the [spec](https://json-schema.org/understanding-json-schema/structuring.html#defs).

---

I've put together a minimal example if you'd like to try it out:

```python
import jsonschema

# note: the $id uses a "namespace" for the project instead of a protocol, which is still valid.
schema = {
    "$schema": "https://json-schema.org/draft-07/schema",
    "$id": "project://test.schema.json",
    "type": "object",
    "properties": {
        "foo": {"$ref": "#/$defs/bar"}
    },
    "$defs": {
        "bar": {"type": "integer"}
    }
}

# In reality there is a dependency between multiple schemas in the same directory or namespace.
# I understand that we can pre-populate a RefResolver with a store of schemas.
store = {"project://test.schema.json": schema}
resolver = jsonschema.RefResolver.from_schema(schema, store=store)

instance = {"foo": 12}
jsonschema.Draft7Validator(schema=schema, resolver=resolver).validate(instance)
```

It should be noted that a workaround exists by appending the current document's id before the fragment. 

```python
schema = {
    "$schema": "https://json-schema.org/draft-07/schema",
    "$id": "project://test.schema.json",
    "type": "object",
    "properties": {
        "foo": {"$ref": "project://test.schema.json/#/$defs/bar"}
    },
    "$defs": {
        "bar": {"type": "integer"}
    }
}
```

This could be done in a preprocessing step to update all local refs within a schema. If the schemas are consumed by multiple apps across language boundaries, this could be a temporary non-invasive solution.

**Comment by ssbarnea** [comment]
@MatthewNielsen27 I still fail to see how can I use an anchored url to validate some data. Everywhere I see a schema being used, is just a JSON data, without and reference.

**Comment by MatthewNielsen27** [comment]
@ssbarnea Yes, you are right. The example I provided was unique to my use case because we preload our schemas into a `RefResolver`.

Looking at the code, I think the primary issue is with the `find_in_subschemas` function. For starters it doesn't even look for `$defs`.

**Comment by ssbarnea** [comment]
I managed to unblock myself by writing a script the splits our combined schema in separated files, ones that are compatible with this library. See https://github.com/ansible/schemas/blob/main/src/update-schemas.py#L65-L86

So basically now users have two options: either use the combined schema with anchors or use standalone ones. I was lucky because I used a relatively flat design. All reused definitions are at the root level `/definitions`.

I still need to add some code that strips definitions that are not needed for current file, but that is just and optimization for reducing the file-size.

**Comment by Julian** [comment]
I still honestly don't really feel I have enough to look into this, and @MatthewNielsen27 I'm not really sure you're talking about the same issue, but it's hard to tell. Whether `id` encountered in a `$def` is supposed to be locatable is a pure spec question (one IIRC the answer to is "yes", so if you observe different behavior and confirm that that's incorrect that's certainly reasonable to file as an issue)

All the rest here is tied up with things not related to the specification or this library, so while I want to help, I can't do that yet until I know what thing is using this library and how -- right now the ansible schemas are not really schemas, they're the VSCode thing. How can I reproduce what you're seeing @ssbarnea, even if the only instructions you have right now involve VSCode or whatever tool is using the files you're talking about?

**Comment by ssbarnea** [comment]
@Julian Hi and thanks for the reply.

All the files from https://github.com/ansible/schemas/tree/main/f are valid JSON Schemas. If you spotted something that makes them not be such, please let me know, I am quite curious. All of them have an `$id` that points to themselves.

What I was trying to explain that is not working with current validator is the ability to use an anchored URL to validate a document. That is something that works with other validators but it does not with our python library.

To give a specific example, `https://github.com/ansible/schemas/blob/main/f/ansible-tasks.json` should perform the same validation as `https://github.com/ansible/schemas/blob/main/f/ansible.json#/definitions/tasks`.

This first one is an URL without an anchor and that works with jsonschema, but the other one does not.

I will try to write a longer article about how to use various tools when developing schemas that should explain it better. I hope the weekend will allow me to do it. I really need to do this because that info is useful for others too. Still, if you want a hands-on experience I would also be happy to show you how I do various stuff on a meet/zoom session, we might both learn new tricks.

**Comment by Julian** [comment]
> All the files from https://github.com/ansible/schemas/tree/main/f are valid JSON Schemas. If you spotted something that makes them not be such, please let me know, I am quite curious. All of them have an $id that points to themselves.

Yes, all of these certainly seem valid at a glance -- what I'm saying is what do you mean by:

> This first one is an URL without an anchor and that works with jsonschema, but the other one does not.

What code are you running to do this, or what tool are you using to do it?

**Comment by MatthewNielsen27** [comment]
@Julian Thanks for the reply. I think you're right - my issue is likely unrelated to this issue.

<img width="986" alt="Screen Shot 2022-05-13 at 4 05 49 PM" src="https://user-images.githubusercontent.com/35040439/168381960-96055d64-8a52-4187-a107-3271d6cf9096.png">

I believed I narrowed down the culprit in the screenshot above ^:
> `_find_in_subschemas(...)` calls `_get_subschemas_cache` which looks for `_SUBSCHEMAS_KEYWORDS` which is only comprised of `("$id", "id", "$anchor", "$dynamicAnchor")`

Should I open a new issue? If the fix is simply adding `"$defs"` to `_SUBSCHEMAS_KEYWORDS` I could open a PR as well.

**Comment by Julian** [comment]
Ideally confirm first what you're referring to is correct under the specification, but yes if you believe it to be a new issue is indeed welcome, with a sample input you expect to work please.

**Comment by MatthewNielsen27** [comment]
@Julian Yes it's correct under the specification https://json-schema.org/understanding-json-schema/structuring.html#defs

**Comment by Julian** [comment]
That section covers `$def` in general, not what it sounded like you were talking about, but let me see an example on the issue.

**Comment by MatthewNielsen27** [comment]
I opened #947 to discuss.

**Comment by Julian** [comment]
Happy to look at this if/when you've got something I can use to reproduce whatever behavior you're referring to!
